src/Ng/SearchIndexFromCatalogCollector.cs

```diff
@@ -101,13 +105,48 @@ public void EnsureCommitted()
                 return;
             }
 
-            _indexWriter.ExpungeDeletes();
-            _indexWriter.Commit(_metadataForNextCommit.ToDictionary());
+            try
+            {
+                var commitTask = CommitIndexAsync();
+
+                // Ensure that the commit finishes within the configured timeout. If the timeout
+                // threshold is reached, an OperationCanceledException will be thrown and will cause
+                // the process to crash. The process MUST be ended as otherwise the commit task may
+                // finish in the background.
+                if (_commitTimeout.HasValue)
+                {
+                    commitTask = commitTask.TimeoutAfter(_commitTimeout.Value);
+                }
+
+                await commitTask;
+            }
+            catch (OperationCanceledException)
+            {
+                _telemetryService.TrackIndexCommitTimeout();
+                _logger.LogError("TIMEOUT Commit timed out!");
```

Comment:
Is there any other info we want to log here, like `NumDocs()`?

Reply:
Good idea. Added